queue/retry.go

```diff
@@ -0,0 +1,272 @@
+	rh.mu.Lock()
+	defer rh.mu.Unlock()
+	if rh.started {
+		return errors.New("cannot set retry handler queue after it's already been started")
+	}
+	rh.queue = q
+	return nil
+}
+
+func (rh *retryHandler) Put(ctx context.Context, j amboy.RetryableJob) error {
+	if j == nil {
+		return errors.New("cannot retry a nil job")
+	}
+	if ctx.Err() != nil {
+		return ctx.Err()
+	}
+
+	rh.mu.Lock()
+	defer rh.mu.Unlock()
+
+	if _, ok := rh.pending[j.ID()]; ok {
+		return errors.Errorf("cannot retry job %s multiple times", j.ID())
+	}
+
+	rh.pending[j.ID()] = j
+
+	return nil
+}
+
+func (rh *retryHandler) Close(ctx context.Context) {
+	rh.mu.Lock()
+	defer rh.mu.Unlock()
+	if rh.cancelWorkers != nil {
+		rh.cancelWorkers()
+	}
+	rh.started = false
+
+	rh.wg.Wait()
+}
+
+func (rh *retryHandler) waitForJob(ctx context.Context) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		default:
+			var j amboy.RetryableJob
+			defer func() {
+				if err := recovery.HandlePanicWithError(recover(), nil, "handling job retry"); err != nil {
+					if j != nil {
+						rh.pending[j.ID()] = j
+					}
+				}
+			}()
+			// TODO (EVG-13540): make this use a channel instead checking in a
+			// no-op loop.
+			j = rh.nextJob()
+			if j == nil {
+				continue
+			}
+			if err := rh.handleJob(ctx, j); err != nil && ctx.Err() == nil {
+				grip.Error(message.WrapError(err, message.Fields{
+					"message":  "could not retry job",
+					"queue_id": rh.queue.ID(),
+					"job_id":   j.ID(),
+				}))
+			}
+
+			// Once the job has been processed (either success or failure),
+			// mark it as processed so it does not attempt to retry again.
+			j.UpdateRetryInfo(amboy.JobRetryOptions{
+				Retryable:  utility.FalsePtr(),
+				NeedsRetry: utility.FalsePtr(),
+			})
+			// TODO (EVG-13540): this has to retry this op until success,
+			// like the theoretical infinite loop in queue.Complete(). It should
+			// also be done in a transaction-like way when the new job is
+			// inserted, so that the swap occurs safely.
+			if err := rh.queue.Save(ctx, j); err != nil {
+				grip.Critical(message.WrapError(err, message.Fields{}))
+			}
+		}
+	}
+}
+
+func (rh *retryHandler) nextJob() amboy.RetryableJob {
+	rh.mu.RLock()
+	defer rh.mu.RUnlock()
+	for id, found := range rh.pending {
+		j := found
+		delete(rh.pending, id)
+		return j
+	}
+	return nil
+}
+
+func (rh *retryHandler) handleJob(ctx context.Context, j amboy.RetryableJob) error {
+	startAt := time.Now()
+	catcher := grip.NewBasicCatcher()
+	timer := time.NewTimer(0)
+	defer timer.Stop()
+	for i := 0; i < rh.opts.MaxRetryAttempts; i++ {
+		if time.Since(startAt) > rh.opts.MaxRetryTime {
+			return errors.Errorf("giving up after %d attempts, %f seconds due to maximum retry time", i, rh.opts.MaxRetryTime.Seconds())
+		}
+
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-timer.C:
+			catcher.Wrapf(rh.tryEnqueueJob(ctx, j), "enqueue retry job attempt %d", i)
```

Comment:
If the job is successfully enqueued, shouldn't this return?

Reply:
Fixed.